### PR TITLE
Integration events' support

### PR DESCRIPTION
**Pull request by Zuravvski**

### CODE REVIEW

File: backend/Shared/HitMeApp.Shared.Infrastructure/Integration/IntegrationEvent.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
+﻿using System;
+
+namespace HitMeApp.Shared.Infrastructure.Integration
+{
+    public abstract class IntegrationEvent
+    {
+        public Guid Id { get; }
+        public DateTime IssuedAt { get; }
```
Comment (abaniuszewicz) [inline backend/Shared/HitMeApp.Shared.Infrastructure/Integration/IntegrationEvent.cs]: Maybe let's just call it `Timestamp`?

File: backend/Shared/HitMeApp.Shared.Infrastructure/Integration/IntegrationEvent.cs [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+        protected IntegrationEvent()
+        {
+            Id = Guid.NewGuid();
+            IssuedAt = DateTime.UtcNow;
+        }
```
Comment (abaniuszewicz) [inline backend/Shared/HitMeApp.Shared.Infrastructure/Integration/IntegrationEvent.cs]: Could this constructor call second one to initialize these properties?

File: backend/Shared/HitMeApp.Shared.Infrastructure/Integration/Memory/InMemoryIntegrationEventBus.cs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+        public Task Publish<TIntegrationEvent>(params TIntegrationEvent[] events) where TIntegrationEvent : IntegrationEvent
+            => Publish(events?.AsEnumerable());
+
+        public Task Publish<TIntegrationEvent>(IEnumerable<TIntegrationEvent> events) where TIntegrationEvent : IntegrationEvent
+            => Task.WhenAll(events.Select(@event => Publish(@event)));
+
+        private Task Publish<TIntegrationEvent>(TIntegrationEvent @event) where TIntegrationEvent : IntegrationEvent
+        {
+            var handlers = _subscriptions.Where(subscription => subscription as IntegrationEventSubscription<TIntegrationEvent> is { })
```
Comment (abaniuszewicz) [inline backend/Shared/HitMeApp.Shared.Infrastructure/Integration/Memory/InMemoryIntegrationEventBus.cs]: Would `subscription is IntegrationEventSubscription<TIntegrationEvent>` work in this case?

File: backend/Shared/HitMeApp.Shared.Infrastructure/Integration/Memory/InMemoryIntegrationEventBus.cs [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+                .Cast<IntegrationEventSubscription<TIntegrationEvent>>()
+                .Select(subscription => subscription.HandlerResolver().Handle(@event))
+                .ToList();
```
Comment (abaniuszewicz) [inline backend/Shared/HitMeApp.Shared.Infrastructure/Integration/Memory/InMemoryIntegrationEventBus.cs]: `Task.WhenAll` accepts `IEnumerable`. We don't have to cast it to `List`.

**Review by abaniuszewicz — CHANGES_REQUESTED** [review]

File: backend/Shared/HitMeApp.Shared.Infrastructure/Integration/IntegrationEvent.cs (hunk 2 above)
Comment (Zuravvski) [reply backend/Shared/HitMeApp.Shared.Infrastructure/Integration/IntegrationEvent.cs]: Oh yeah, it sure can.

File: backend/Shared/HitMeApp.Shared.Infrastructure/Integration/IntegrationEvent.cs (hunk 1 above)
Comment (Zuravvski) [reply backend/Shared/HitMeApp.Shared.Infrastructure/Integration/IntegrationEvent.cs]: Yup, let's do it

File: backend/Shared/HitMeApp.Shared.Infrastructure/Integration/Memory/InMemoryIntegrationEventBus.cs (hunk 3 above)
Comment (abaniuszewicz) [reply backend/Shared/HitMeApp.Shared.Infrastructure/Integration/Memory/InMemoryIntegrationEventBus.cs]: Alternatively, with use of `as` we could chain it like so
```csharp
var handlers = _subscriptions.Select(subscription => subscription as IntegrationEventSubscription<TIntegrationEvent>)
.Where(subscription => subscription != null)
.Select(subscription => subscription.HandlerResolver().Handle(@event));
```

As right now we're making cast twice (with use of `as` operator and then using `Cast` method).

File: backend/Shared/HitMeApp.Shared.Infrastructure/Integration/Memory/InMemoryIntegrationEventBus.cs (hunk 4 above)
Comment (Zuravvski) [reply backend/Shared/HitMeApp.Shared.Infrastructure/Integration/Memory/InMemoryIntegrationEventBus.cs]: The signature was different before. Right now it makes no sense to keep it that way. Corrected.

File: backend/Shared/HitMeApp.Shared.Infrastructure/Integration/Memory/InMemoryIntegrationEventBus.cs (hunk 3 above)
Comment (Zuravvski) [reply backend/Shared/HitMeApp.Shared.Infrastructure/Integration/Memory/InMemoryIntegrationEventBus.cs]: I dunno about the corner cases, but I did similar things in the past and they worked quite well.
As a backup plan I store `Type`  in the subscription so we can do it differently.  Let's see about that after we introduce a second module to integrate with =)

Comment (Zuravvski) [reply backend/Shared/HitMeApp.Shared.Infrastructure/Integration/Memory/InMemoryIntegrationEventBus.cs]: Double casting was indeed unnecessary. I simplified it according to your solution.

**Review by abaniuszewicz — APPROVED** [review]

**Pull request merged**